File: compleat/python/without_h/gen_training_data.py
# -*- coding: utf-8 -*- 
import numpy as np
import matplotlib.pyplot as plt
q, L = 3, 2 
X = np.random.choice(q,L) # Chose L-variables from q-state .
J = np.zeros((L*L, q*q))
J0 = 0.3  

T_equil = int(1e3) 
T_relax = int(1e2) 
N_sampling = int(1e4)

def J_generalized_potts():
    global J
    for i in range(L):
        for j in range(i+1,L):
            for a in range(q):
                for b in range(q):
                    r = np.random.normal() * J0 
                    J[i*L+j][a*q+b] = r  
                    J[j*L+i][b*q+a] = r 
    fname = "J_data_L"+str(L)+".dat"
    f = open(fname, "w")
    #J_vec = np.reshape( np.copy(J), q**2*L**2)
    for i in range(L):
        for j in range(L):
            for a in range(q):
                for b in range(q):
                    f.write(str(J[i*L+j][a*q+b])+"\n")
    #for l in range(len(J_vec)):
    #    f.write(str(J_vec[l])+" \n")
    #print "J=\n",J
    f.close()

def E_local(i):
    a = X[i]
    E_i = 0.0
    for j in range(L): # NOTE: J[i*L+j][xx] = 0
        b = X[j] 
        E_i += -J[i*L+j][a*q+b] 
    return E_i

def E():
    E_tot = .0
    for i in range(L):
        E_tot = E_local(i)
    return E_tot / 2.0

def Metropolis(i):
    global X 
    E_i = E_local(i)
    Xi_save = X[i]
    X[i] = (X[i]+np.random.choice(q-1)+1) % q  
    E_i_trial = E_local(i)
    dE = E_i_trial - E_i
    w = np.exp(-dE)
    accepted = False
    if(np.random.uniform() < w):
        accepted = True
    else:
        X[i] = Xi_save
    return accepted

def MonteCarlo_sweep():
    accepted = .0
    for l in range(L):
        i = np.random.choice(L)
        if(Metropolis(i)):
            accepted +=1
    return accepted

if __name__ == "__main__":
    J_generalized_potts()
    fname  = "test_training_data_L"+str(L)+".dat"
    fout = open(fname,"w")
    for t in range(T_equil):
        MonteCarlo_sweep()
    for t in range(T_relax*N_sampling):
        MonteCarlo_sweep()
        if(t%T_relax == 0):
            for i in range(L):
                fout.write(str(X[i])+" ")
            fout.write("\n")
    fout.close() 
    #Engy /= N_sampling 
    #print "Engy=", Engy

File: compleat/c++/without_Hidden/without_h/gen_training_data.cpp
// Generative model of potts. 
#include <cmath>
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <iomanip>
using namespace std;

int q=3, L=4;
double J0 = 1.0;

/*****************************/
vector<int> X(L,0);
vector < vector<double> > J( L*L, vector<double>(q*q,0) );  
/*****************************/
int T_equil = 1e3;
int T_interval = 1e2;
int N_sampling = 1e4;
/*****************************/
inline double std_rand()
{
    return rand() / (RAND_MAX + 1.0);
}

void J_generalized_potts(){
	double r;
	for(int i=0;i<L;i++){
	for(int j=i+1;j<L;j++){
	//for(int j=0;j<L;j++){
		for(int a=0;a<q;++a){
		for(int b=0;b<q;++b){
			r = std_rand();
		//	J[i*L+j][a*q+b] = count;
			J[i*L+j][a*q+b] = r;
			J[j*L+i][b*q+a] = r;
		}}}}
	string fname = "J_data_L"+to_string(L)+"_c++.dat";
	ofstream file(fname);
	for(int i=0;i<L;i++){
	for(int j=0;j<L;j++){
		for(int a=0;a<q;++a){
		for(int b=0;b<q;++b){
			file << J[i*L+j][a*q+b] << endl; 
		}}}}
	file.close();
}
double E_local(int i){
	int a = X[i],b;
	double E_i = 0.0;
	for(int j=0;j<L;++j){
		b = X[j];
		E_i += -J[i*L+j][a*q+b];	
	}
	return E_i;	
}

double E(){
	double E_tot = 0.0;
	for(int i=0;i<L;++i){
		E_tot = E_local(i);
	}
	return E_tot / 2.0;
}

bool Metropolis(int i){
	double E_i, E_i_trial, dE, w;
	int Xi_save;
	bool accepted;
	
	E_i= E_local(i);
	Xi_save = X[i];
	X[i] = (X[i]+int(std_rand()*q+1)) % q;
	E_i_trial = E_local(i);
	dE = E_i_trial - E_i;
	w =exp(-dE);
       	accepted = false;
	if(w > std_rand())
       		accepted = true;
	else
		X[i] = Xi_save;
	return accepted;
}

int MonteCarlo_sweep(){
	double accepted;
	int i,l;
	for(l=0;l<L;++l){
		i = int(L*std_rand());
		if(Metropolis(i))
			accepted += 1;
	}
	return accepted;
}

int main(){
	J_generalized_potts();
	string fname = "test_training_data_L" + to_string(L) + "_c++.dat";
	ofstream file(fname);
	int t;
	t = 0;
	while(t<T_equil){
		MonteCarlo_sweep();	
		t+=1;
	}
	t = 0;
	while(t<T_interval*N_sampling){
		MonteCarlo_sweep();
		if(t%T_interval == 0){
		
		for(int i=0;i<L;++i){
			file << X[i] << " ";	
		}
			file << endl;
	}
		t+=1;
	}
	file.close();
}


